File: CodingTest/Coding_Test/BAEKJOON/2812.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        int n=sc.nextInt();
        int k=sc.nextInt();
        String[] num=sc.next().split("");
        Stack<String> stack=new Stack<>();
        int lastL=num.length;
        int curL=0;
        for(int i=0;i<num.length;i++){
            if(stack.isEmpty()){
                stack.push(num[i]);
                lastL--;
                curL++;
                continue;
            }
            if(lastL+curL<=n-k||Integer.parseInt(stack.peek())>=Integer.parseInt(num[i])){
                if(curL<n-k){
                    stack.push(num[i]);
                    curL++;
                    lastL--;
                }
            }else{
                while(!stack.isEmpty()&&lastL+curL>n-k&&Integer.parseInt(stack.peek())<Integer.parseInt(num[i])){
                    stack.pop();
                    curL--;
                }
                stack.push(num[i]);
                curL++;
                lastL--;
            }
        }
        // System.out.println(stack.toString());
        StringBuilder sb=new StringBuilder();
        for(int i=0;i<stack.size();i++){
            sb.append(stack.get(i));
            // System.out.println(stack.get(i));
        }
        System.out.println(sb.toString());
    }
}

File: CodingTest/Coding_Test/BAEKJOON/2812.py
from collections import deque

n,k=map(int,input().split())
num=list(input())
lastL=len(num)
curL=0
answer=deque([])
for i in range(len(num)):
    if len(answer)==0:
        answer.append(num[i])
        lastL-=1
        curL+=1
        continue
    if lastL+curL<=n-k or int(answer[-1])>=int(num[i]):
        if curL<n-k:
            answer.append(num[i])
            curL+=1
            lastL-=1
    else:
        deleteCnt=0
        while len(answer)!=0 and lastL+curL>n-k and int(answer[-1])<int(num[i]):
            answer.pop()
            curL-=1
        answer.append(num[i])
        curL+=1
        lastL-=1

print("".join(answer))